### DISCUSSION TITLE
How does this format the file?

### DISCUSSION THREAD

**Discussion by cvkmohan** [elixir-lsp/vscode-elixir-ls#224] [Q&A] [answered]
When we press ```Ctrl+Shift+F``` or ```editor.formatOnSave``` is true - does this call ```mix.format```? If I add a plugin in the ```formatter.exs``` the sigil is not getting formatted on save.  
How to make it work?

**Comment by axelson** [comment, marked as answer]
Hi! ElixirLS doesn't call `mix format` directly for two main reasons:
- It would be very slow because we'd have to wait for the BEAM to startup
- We want to support undo/redo via LSP

Here's the code that does the formatting: https://github.com/elixir-lsp/elixir-ls/blob/1f1922db65f0ee7b1d2e754dffdba487ae3f2f01/apps/language_server/lib/language_server/providers/formatting.ex

ElixirLS hasn't added support for this new Elixir 1.13 feature yet so I've create a feature request for this: https://github.com/elixir-lsp/elixir-ls/issues/660

**Reply by msaraiva** [reply]
Hey @axelson!

Does the VM running `ElixirLS.LanguageServer.Providers.Formatting` have the project's beams loaded? 

If so, it could still properly format the code with plugins and stuff if you pass the result of [formatter_opts_for_file/2](https://github.com/elixir-lang/elixir/blob/v1.13.2/lib/mix/lib/mix/tasks/format.ex#L231) to the `Code.format_string!/2` call.

For Elixir >= `1.13`, you could even skip `Code.format_string!/2` and use the formatter returned by [formatter_for_file/2](https://github.com/elixir-lang/elixir/blob/v1.13.2/lib/mix/lib/mix/tasks/format.ex#L217) directly.

**Reply by zachdaniel** [reply]
I'm trying to figure out why it doesn't already work, since it does get the formatter opts from the mix task logic:
https://github.com/elixir-lsp/elixir-ls/blob/1f1922db65f0ee7b1d2e754dffdba487ae3f2f01/apps/language_server/lib/language_server/source_file.ex#L336

**Reply by msaraiva** [reply]
@zachdaniel maybe the plugin modules are available but not loaded? We could try to go through the list of plugins and call `Code.ensure_loaded/1` for each one of them. In ElixirSense, all modules from the project and dependencies are available, otherwise we wouldn't be able to introspect information to provide suggestions, etc. so I assume they should be available (but maybe not loaded) when calling the formatter too. In case they're not, we could move the formatter logic to ElixirSense, but only if we can confirm they don't run on the same VM instance.

**Reply by zachdaniel** [reply]
There is an interesting question of formatter plugins and what they can do here, also. Specifically, I have a formatter plugin that overrides `.ex` files for Ash resources. There are a couple of issues with that kind of thing:

1. you can only have one of them potentially. I remove the ash formatter from the plugins list and then call the formatter again, so maybe the "next in line" that overrides .ex files would get triggered? I imagine they group them by file extension/sigil so you don't end up with clashes... 🤷🏻 

2. what happens in cases where running the formatter *without* your plugin would cause a *reversal* of the formatting? This isn't the case right now (and I can almost certainly keep it that way with further additions to the ash formatter). But what about having some kind of special formatter plugin that allows things that the built in formatter *wouldn't* allow? Letting the formatter run without plugins ever would mean that you could put things in a bad state, maybe even in a state such that the file is changed and neither the base formatter nor the plugin would put it back the way it should have been (the way that adding parens is not reversable by adding something to `locals_without_parens`).

So it stands to reason that, if there are any plugins at all, we shouldn't ever format without all plugins being loaded. For some of my projects, compile times are non trivial. In the case of a partial recompile, does elixir ls "keep working" or does everything it does wait for the recompile?

Sorry for the ramble/wall of text. Hopefully it is just as simple as adding a `Code.ensure_loaded/1` :D

**Reply by msaraiva** [reply]
@zachdaniel those are valid questions but I believe addressing them is probably out of our scope here. I mean, those concerns are also valid when running `mix format`, right? If so, there's not much we can do about it and they should be handled at the Elixir core level. So I believe if we manage to keep the same behavior of `mix format`, we'll be just fine.

**Reply by zachdaniel** [reply]
Yeah, my point only is that 

> It would be very slow because we'd have to wait for the BEAM to startup

May not be a concern because now that plugins are supported, we basically *must* wait for those modules to be compiled ?

**Reply by zachdaniel** [reply]
Otherwise we risk making unfixable (except manually) changes to codebases that use plugins.

**Reply by msaraiva** [reply]
@zachdaniel I didn't mean we should call `mix format`. That would certainly be too slow. What I meant is that there must be already a VM up and running for ElixirLS and since the plugin modules all belong to dependencies, they should've already been compiled. If so, we could load those modules then use `Code.format_string`, passing the same options that `mix format` would use. As far as I know that's all `mix format` does. Unless, of course, I'm missing something here :)

**Reply by zachdaniel** [reply]
Understood 👍🏻  There is warning output if a plugin is passed but isn't loaded IIRC, so maybe we can see in elixirls output if that is what happening.
